darwin/future/core/types/query.py

```diff
@@ -106,47 +110,38 @@ def __len__(self) -> int:
             return 0
         return len(self.filters)
 
-    def __iter__(self) -> Query[T]:
+    def __iter__(self) -> Query[T, R]:
         self.n = 0
         return self
 
-    def __next__(self) -> QueryFilter:
-        if self.filters is None:
-            self.filters = []
-        if self.n < len(self.filters):
-            result = self.filters[self.n]
+    def __next__(self) -> R:
+        if self.results is None:
+            self.results = list(self.collect())
+        if self.n < len(self.results):
+            result = self.results[self.n]
             self.n += 1
             return result
         else:
             raise StopIteration
+        
+    def __getitem__(self, index: int) -> R:
```

Comment:
Are our keys definitely always integers?  `__setitem__` and `__getitem__` also take strings as native, but I can't immediately think whether we have any issue with that.

Reply:
as in just implicitly converting a string to an int here, like query['6'] or would there be some other meaningful interaction with a string as an index that we'd want? Like query['<name>'] returning the dataset name for instance? I would worry that has too much overlap with like query.where('name=...') and we'd just be maintaining multiple sets of functionality that do the same job. IMO better to just have and maintain the one.